File: templatemethod/java/Pizza.java
package templatemethod.java;

public class Pizza {
    public void templateMethod() {
        prepareDough();
        addSauce();
        addToppings();
        bake();
        slice();
        box();
    }

    protected void prepareDough() {
        System.out.println("Preparing the dough");
    }

    protected void addSauce() {
        System.out.println("Adding sauce");
    }

    protected void addToppings() {
        System.out.println("Adding toppings");
    }

    protected void bake() {
        System.out.println("Baking the pizza");
    }

    protected void slice() {
        System.out.println("Slicing the pizza");
    }

    protected void box() {
        System.out.println("Boxing the pizza");
    }
}

File: templatemethod/rust/templatemethod/src/pizza.rs
pub trait Pizza {
    // The Template Method: Defines the skeleton of the algorithm
    fn make_pizza(&self) {
        self.prepare_dough();
        self.add_sauce();
        self.add_toppings(); // Step to be overridden
        self.bake();
    }

    fn prepare_dough(&self) {
        println!("Preparing standard thin crust dough.");
    }

    fn add_sauce(&self) {
        println!("Adding tomato sauce.");
    }

    // This method is intended to be overridden by "subclasses"
    fn add_toppings(&self);

    fn bake(&self) {
        println!("Baking for 15 minutes at 400°F.");
    }
}
